Creating a toolchain package docs (#3625)
conan-io/docs: tutorial/creating_packages/other_types_of_packages/toolchain_packages.rst

.. _tutorial_other_toolchain_packages:

> Creating a Conan package for a toolchain

Author: Yes, I had the same feeling... another idea would be to move it to an example, and also move the part that uses the toolchain to the tutorial that uses the profile linking to the example...

> Creating a Conan package for a toolchain

Reviewer: Yes, that would make sense, having the "usage" of the toolchain in the tutorial, and the full explanation of the recipe, the ``setting_target``, etc, in its dedicated section in the examples.

> Creating a Conan package for a toolchain

Author: great! I'll move it

> Creating a Conan package for a toolchain

Author: In the end I moved it to an example, but I left the part of using it in the example and linked it in the tutorial to not extend the tutorial too much

            for dir_name in dirs_to_copy:
                copy(self, pattern=f"{dir_name}/*", src=self.build_folder, dst=self.package_folder, keep_path=True)
            copy(self, "LICENSE", src=self.build_folder, dst=os.path.join(self.package_folder, "licenses"), keep_path=False)

        def package_info(self):
            toolchain, _ = self._get_toolchain(self.settings_target.arch)
            self.cpp_info.bindirs.append(os.path.join(self.package_folder, toolchain, "bin"))

            self.conf_info.define("tools.build:compiler_executables", {
                "c":   f"{toolchain}-gcc",
                "cpp": f"{toolchain}-g++",
                "asm": f"{toolchain}-as"
            })

Validating the toolchain package: settings, settings_build and settings_target
------------------------------------------------------------------------------

As you may recall, the :ref:`validate() method<reference_conanfile_methods_validate>` is
used to indicate that a package is not compatible with certain configurations. As
mentioned earlier, we are limiting the usage of this package to a *Linux x86_64* platform
for cross-compiling to a *Linux ARM* target, supporting both 32-bit and 64-bit

> architectures. Let's check how we incorporate this information into the ``validate()``

Reviewer: I think this section will greatly benefit from a graphviz representation of the build/host/target contexts

- Add directories containing toolchain tools that may be required during compilation. The
  toolchain we download will store its tools in both ``bin`` and
  ``<toolchain_triplet>/bin``. Since ``self.cpp_info.bindirs`` defaults to ``bin``, we
  only need to add the directory specific to the triplet. Note that it's not necessary to
  define environment information to add these directories to the ``PATH``, as Conan will
  manage this through the :ref:`VirtualRunEnv<conan_tools_env_virtualrunenv>`.

- We define the ``tools.build:compiler_executables`` configuration. This configuration
  will be considered in several generators, like
  :ref:`CMakeToolchain<conan_tools_cmaketoolchain>`,
  :ref:`MesonToolchain<conan_tools_meson_mesontoolchain>`, or
  :ref:`AutotoolsToolchain<conan_tools_gnu_autotoolstoolchain>`, to direct to the
  appropriate compiler binaries.

Testing the Conan toolchain package
-----------------------------------

We also added a simple *test_package* to test the toolchain:

.. code-block:: python
    :caption: test_package/conanfile.py

    import os
    from io import StringIO

    from conan import ConanFile
    from conan.tools.cmake import CMake, cmake_layout

>     class TestPackgeConan(ConanFile):
>         settings = "os", "arch", "compiler", "build_type"
>         generators = "CMakeToolchain", "VirtualBuildEnv"
>         test_type = "explicit"
> 

Author: ```suggestion
    class TestPackageConan(ConanFile):
        settings = "os", "arch", "compiler", "build_type"
        generators = "CMakeToolchain", "VirtualBuildEnv"

```

        def build_requirements(self):
            self.tool_requires(self.tested_reference_str)

        def layout(self):
            cmake_layout(self)

        def build(self):
            cmake = CMake(self)
            cmake.configure()
            cmake.build()

        def test(self):
            if self.settings.arch in ["armv6", "armv7", "armv7hf"]:
                toolchain = "arm-none-linux-gnueabihf"
            else:
                toolchain = "aarch64-none-linux-gnu"
            self.run(f"{toolchain}-gcc --version")
            test_file = os.path.join(self.cpp.build.bindirs[0], "test_package")
            stdout = StringIO()
            self.run(f"file {test_file}", stdout=stdout)
            if toolchain == "aarch64-none-linux-gnu":
                assert "ELF 64-bit" in stdout.getvalue()
            else:
                assert "ELF 32-bit" in stdout.getvalue()

> This test package ensures that the toolchain is functional building a minimal *hello world*

Reviewer 2: ```suggestion
This test package ensures that the toolchain is functional, building a minimal *hello world*
```

is intended as a build requirement, situating it within the build context. Consequently,
`settings` match those from the build profile, while `settings_target` aligns with the
host profile's settings.

With the toolchain package prepared, we proceed to build an actual application. This will
be the same application previously cross-compiled in the
:ref:`consuming_packages_cross_building_with_conan` section. However, this time, we
incorporate the toolchain package as a dependency within the host profile. This ensures
the toolchain is used to build the application and all its dependencies

.. code-block:: bash

    $ cd .. && cd consumer
    $ conan install . -pr:b=default -pr:h=../profiles/raspberry-64 -pr:h=../profiles/arm-toolchain --build missing
    $ cd build && cmake .. -DCMAKE_TOOLCHAIN_FILE=Release/generators/conan_toolchain.cmake -DCMAKE_BUILD_TYPE=Release
    $ cmake --build .
    $ file compressor 
    compressor: ELF 64-bit LSB executable, ARM aarch64, version 1 (SYSV), dynamically
    linked, interpreter /lib/ld-linux-aarch64.so.1, for GNU/Linux 3.7.0, with debug_info,
    not stripped

> We composed the already existing profile with other profile that just has

Reviewer 2: ```suggestion
We composed the already existing profile with another profile called ``arm-toolchain`` that just has
```
